src/main/java/digital/slovensko/autogram/core/Autogram.java

```diff
@@ -60,27 +67,83 @@ public void sign(SigningJob job, SigningKey signingKey) {
                 job.signWithKeyAndRespond(signingKey);
                 ui.onUIThreadDo(() -> ui.onSigningSuccess(job));
             } catch (DSSException e) {
-                ui.onUIThreadDo(() -> ui.onSigningFailed(AutogramException.createFromDSSException(e)));
+                onSigningFailed(AutogramException.createFromDSSException(e));
             } catch (IllegalArgumentException e) {
-                ui.onUIThreadDo(() -> ui.onSigningFailed(AutogramException.createFromIllegalArgumentException(e)));
+                onSigningFailed(AutogramException.createFromIllegalArgumentException(e));
             } catch (Exception e) {
-                ui.onUIThreadDo(() -> ui.onSigningFailed(new UnrecognizedException(e)));
+                onSigningFailed(new UnrecognizedException(e));
             }
         });
     }
 
+    /**
+     * Starts a batch - ask user - get signing key - start batch - return batch ID
+     * 
+     * @param totalNumberOfDocuments - expected number of documents to be signed
+     * @param responder              - callback for http response
+     */
+    public void batchStart(int totalNumberOfDocuments, BatchResponder responder) {
+        if (batch != null && !batch.isEnded())
+            throw new BatchConflictException("Iné hromadné podpisovanie už prebieha");
+        batch = new Batch(totalNumberOfDocuments);
+
+        var startBatchTask = new AutogramBatchStartCallback(batch, responder);
+
+        ui.onUIThreadDo(() -> {
+            ui.startBatch(batch, this, startBatchTask);
+        });
+    }
+
+    /**
+     * Sign a single document
+     * 
+     * @param job
+     * @param batchId - current batch ID, used to authenticate the request
+     */
+    public void batchSign(SigningJob job, String batchId) {
+        if (batch == null)
+            throw new BatchNotStartedException();
+
+        batch.addJob(batchId, job);
+
+        if (Platform.isFxApplicationThread()) {
+            ui.onWorkThreadDo(() -> {
+                ui.signBatch(job, batch.getSigningKey());
+            });
+        } else {
+            ui.signBatch(job, batch.getSigningKey());
+        }
```

Comment:
Toto by to uz nemalo byt ked to robi ten onWorkThreadDo sam nie?

Reply:
ano, good catch